lib/storage/metadata/mongoclient/ListRecordStream.js

```diff
@@ -1,32 +1,42 @@
 const stream = require('stream');
 
 class ListRecordStream extends stream.Transform {
-    constructor(logger, lastEndID) {
+    constructor(logger, lastSavedID, latestOplogID) {
         super({ objectMode: true });
         this._logger = logger;
-        this._lastEndID = lastEndID;
-        this._lastTs = 0;
+        this._lastSavedID = lastSavedID;
+        this._latestOplogID = latestOplogID;
         this._lastUniqID = null;
         // this._unpublishedListing is true once we pass the oplog
-        // that has the start seq timestamp and uniqID 'h'
+        // record that has the same uniqID 'h' than last saved. If we
+        // don't find it (e.g. log rolled over before populator could
```

Comment:
The oplog is a "capped collection", MongoDB uses it internally as such, here's the MongoDB doc description:

> Capped collections are fixed-size collections that support high-throughput operations that insert and retrieve documents based on insertion order. Capped collections work in a way similar to circular buffers: once a collection fills its allocated space, it makes room for new documents by overwriting the oldest documents in the collection.

So the log rolls when it runs out of space (so space-based rather than time-based), and it indeed means oldest entries will get overwritten by newer ones.

The default sizing seems to be 5% of available storage space, with a minimum of 990MB (https://docs.mongodb.com/v3.4/core/replica-set-oplog/#replica-set-oplog-sizing)

For this reason at least, we will keep the "sweeper" in case old entries are overwritten before they are being processed, either as a manually-triggered task or as an automatically scheduled periodic task.

Reply:
nice, thank you for the info!